File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        check(root);
        return max;
    }

private:
    int max = INT_MIN;

    int check(TreeNode* root) {
        if (root == nullptr) return 0;

        int maxL = check(root->left);
        int maxR = check(root->right);

        int addL = maxL > 0 ? maxL : 0;
        int addR = maxR > 0 ? maxR : 0;

        int imTheTurn = root->val + addL + addR;

        if (imTheTurn > max)
            max = imTheTurn;

        int imNotTheTurn = addL > addR ? root->val + addL : root->val + addR;

        return imNotTheTurn; // someone else is
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, m):
            good = 0
            if node is None:
                return 0
            if node.val >= m:
                good += 1
            good += dfs(node.left, max(node.val, m))
            good += dfs(node.right, max(node.val, m))
            return good

        return dfs(root, float("-inf"))
